### PR TITLE
feat(ir): add missing math operators

### LINKED ISSUE #2665: [Feature Request] Comprehensive Operator Library — 16 composite operators + 1 missing primitive (gather) for NPU acceleration
### Summary

PyPTO currently provides only low-level Tile primitives (similar to Triton's `tl.load/tl.store/tl.dot`), with no high-level operator library. We developed 6 neural network models (ViT-Frame, ViT-MultiView, ViT-Temporal, BigTransformer-MoE, Planner, VAE-Encode) using the Tile-level DSL and found that **all common neural network operators must be hand-composed from primitives**. This is extremely labor-intensive and error-prone.

We request:
1. **A new `gather` primitive** — the only missing primitive blocking conv and interpolation
2. **16 composite operators** as a standard operator library — 10 we already implemented (please adopt/maintain), 6 we couldn't implement due to missing primitives or edge-case constraints

---

### Background

We built 6 production-quality networks on PyPTO using `@pl.jit` + `pl.load/store/matmul` Tile-level DSL, all verified on a2a3sim with golden alignment (threshold=0.005, FP16 compute + FP32 accumulate):

| Network | Architecture | JIT Coverage | Golden max_abs |
|---------|-------------|-------------|----------------|
| ViT-Frame | PatchEmbed - 2xVisionBlock - PatchMerger | 100% | 0.017578 |
| ViT-MultiView | Same as Frame, no spatial merge | 100% | 0.004883 |
| ViT-Temporal | Cross-attention (Q!=KV) + KV precompute | 100% | 0.002930 |
| BigTransformer | MoE DecoderLayer x2 (4 experts, top-2) | 100% | 0.004756 |
| Planner | TimeEmbed - GridSample - CrossAttn - Decoder | ~60% | 0.001648 |
| VAE-Encode | CausalConv3d - 4-level DownResBlock - Middle | ~90% | 0.000122 |

---

### Current Framework Primitives (sufficient for composition)

These are the primitives PyPTO already provides, which we used to build all composite operators:

| Category | Primitives |
|----------|-----------|
| Data movement | `pl.load`, `pl.store`, `pl.create_tile`, `pl.create_tensor`, `pl.assemble` |
| Element-wise | `pl.mul`, `pl.add`, `pl.sub`, `pl.div`, `pl.neg`, `pl.exp`, `pl.recip`, `pl.sqrt`, `pl.abs`, `pl.maximum`, `pl.minimum` |
| Row reduction | `pl.row_max`, `pl.row_sum` |
| Row broadcast | `pl.row_expand_div`, `pl.row_expand_sub`, `pl.row_expand_expdif`, `pl.row_expand_mul` |
| Column broadcast | `pl.col_expand_mul`, `pl.col_expand` |
| Matrix multiply | `pl.matmul`, `pl.matmul_acc` |
| Shape | `pl.reshape`, `pl.transpose` |
| Control flow | `pl.range`, `pl.parallel`, `pl.pipeline`, `pl.at` |

---

### Request 1: New Primitive — `gather` (Dynamic Indexing)

**This is the single most critical missing primitive.** Without it, convolution and bilinear interpolation cannot be composed from existing primitives.

```python
# Proposed API
gathered = pl.gather(src_tensor, indices)  # indices can be runtime-computed
```

**Why it's needed:**

| Operator | Why gather is required |
|----------|----------------------|
| Conv3d | Sliding window requires dynamic offset indexing into input feature map |
| Conv2d (stride-2) | Same sliding window + stride subsampling |
| GridSample | Bilinear interpolation requires `input[floor(x), floor(y)]` where x, y are runtime-computed coordinates |

**Current workaround:** Fall back to `torch` (CPU), which defeats the purpose of NPU acceleration.

**Impact:** 2 of our 6 networks (Planner, VAE-Encode) have torch fallbacks purely because of this missing primitive. Adding `gather` would enable full NPU coverage for all 6 networks.

---

#### Category A: 10 Operators We Already Implemented (Requesting Framework Adoption)

We hand-composed these from primitives. They are verified and working. **We request the framework team to adopt, optimize, and maintain them as a standard library**, rather than each user reinventing the wheel.

| # | Operator | Composition (primitives used) | Input Shape | Verified max_diff | Used By |
|---|----------|-------------------------------|-------------|-------------------|---------|
| 1 | `softmax` | `row_max -> row_expand_expdif -> row_sum -> row_expand_div` | [128, 128] | 2.98e-08 | ViT x3, BigTransformer, Planner, VAE |
| 2 | `sigmoid` | `neg -> exp -> add(1) -> recip` | [192, 256] | 1.19e-07 | BigTransformer, Planner, VAE |
| 3 | `silu` | `sigmoid -> mul` (x * sigmoid(x)) | [192, 256] | 2.38e-07 | Planner, VAE |
| 4 | `gelu_tanh` | `pow -> add -> tanh approx -> mul -> add` | [192, 2048] | 4.77e-07 | (library) |
| 5 | `fast_gelu_v2` | Piecewise polynomial: `mul -> add -> sigmoid -> mul` | [192, 2048] | 0.0 (exact) | ViT x3 |
| 6 | `layer_norm` | `row_sum(mean) -> mul+row_sum(var) -> rsqrt -> sub -> mul` | [144, 512] | 9.54e-07 | ViT x3 |
| 7 | `rms_norm` | `mul(pow2) -> row_sum -> rsqrt -> mul` | [144, 512] | 1.91e-06 | BigTransformer, Planner, VAE |
| 8 | `element_mul` | `pl.row_expand_mul` (element-wise multiply) | [192, 256] | verified | BigTransformer (MoE SwiGLU) |
| 9 | `group_norm` | Per-group: `mean -> var -> rsqrt -> sub -> mul` (grouped layer_norm) | [144, 512] | verified | VAE-Encode |
| 10 | `topk` (top-2) | `row_max -> arithmetic mask -> row_max` (tournament method) | [128, 4] | verified | BigTransformer (MoE routing) |

**Key implementation challenges we solved (so the framework team doesn't have to):**

1. **Vec buffer limit (188,416 bytes):** Element-wise tiles must use [64,128] or [96,128]; norm tiles must use [16,512] or [24,512]. Larger tiles overflow.
2. **FP32 accumulation in FP16 networks:** Softmax pre-scale matmul must use FP32 — FP16 rounding flips near-tied softmax logits, causing >5000x error.
3. **Small dimension padding:** When M < 16 (fractal minimum), zero-pad rows to 16 and set padded softmax scores to `-inf`.
4. **NZ (fractal) layout constraints:** `pl.reshape` does NOT do flat reinterpretation under NZ layout — manual slice+assemble or transpose chains are required.
5. **topk without gather:** Used arithmetic mask tournament (`max -> mask with -inf -> max`), avoiding any dynamic indexing.

#### Category B: 6 Operators We Could NOT Implement (Blocked by Missing Primitives or Edge Cases)

| # | Operator | Used By | Blocker | Proposed Solution |
|---|----------|---------|---------|-------------------|
| 11 | `conv3d` (causal) | VAE-Encode | Needs `gather` for sliding window | Add `gather` primitive, then im2col + matmul |
| 12 | `conv2d` (stride-2) | VAE-Encode | Same as above | Add `gather`, then im2col + matmul + stride subsample |
| 13 | `grid_sample` (bilinear) | Planner | Needs `gather` for dynamic coordinate indexing | Add `gather`, then 8x index + 6x mul + 2x add |
| 14 | `relu` | Planner | Trivially composable (`pl.maximum(x, 0)`) but not yet in library | Add as 1-line wrapper |
| 15 | `output_normalize` (pow/mean/sqrt/clamp) | VAE-Encode | Composable (`mul -> row_sum -> rsqrt -> maximum/minimum`) but not yet written | Add as composite operator |
| 16 | `sinusoidal_pos_emb` (sin/cos/cat) | Planner | Needs `sin`/`cos` primitives (not in current framework) | Add `sin`/`cos` Vec primitives, or precompute on host |

---

### Request 3: Small-Dimension Support (Fractal Alignment)

Several operators fail when dimensions are smaller than the fractal minimum (16). This is a recurring pain point:

| Case | Dimension | Problem | Current Workaround |
|------|-----------|---------|-------------------|
| Planner output head | N=1, N=2 | `matmul` requires N>=16 | Pad weight to N=16, slice output |
| Planner attention | M=4 | `row_max`/`row_sum` tile expects M>=16 | Zero-pad rows to 16, mask with `-inf` |
| Planner traj embed | N=2 | Same as above | Same padding approach |
| BigTransformer router | N=4 | Same | Fall back to `torch.nn.functional.linear` |

**Request:** Either (a) relax the fractal minimum to support N/M >= 1, or (b) provide a built-in padding+slice utility so users don't hand-code it every time.

---

### Proposed Priority

| Priority | Item | Impact |
|----------|------|--------|
| P0 | `gather` primitive | Unblocks conv3d, conv2d, grid_sample -> full NPU coverage for all 6 networks |
| P1 | 10 composite operators (Category A) | Eliminates reinvention; we provide verified reference implementations |
| P2 | `sin`/`cos` primitives | Unblocks positional embedding |
| P2 | Small-dimension support | Eliminates padding boilerplate |
| P3 | `relu`, `output_normalize` wrappers | Trivial but should be in standard library |

---

### Reference Implementation

Our verified implementations are in:

```
examples/models/interaction/base_ops.py          # 9 composite operators (770 lines)
examples/models/interaction/big_transformer.py   # topk tournament (inline)
```

All operators pass a2a3sim simulation with golden alignment. We're happy to contribute these as a PR if the team agrees on the API design.

---

### Environment

- PyPTO: latest (git, untracked `examples/models/interaction/`)
- CANN: 9.2.0
- Simulator: a2a3sim (CPU mode, no Ascend hardware)
- Python: 3.11 (ptoas 0.60 cp311 ABI requirement)

### Proposed API / Behavior

_No response_

### Alternatives Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

- cover the four standalone math operators requested by #2665:
  - `pow` — add tensor, tile, and unified APIs
  - `mean` — add tensor, tile, and unified APIs
  - `sqrt` — reuse the existing operator and add dedicated A2/A3 ST coverage
  - `clamp` — add tensor, tile, and unified APIs
- lower the new composite APIs through existing A2/A3 primitives with FP32 intermediates where computation is required
- document the complete `LowerCompositeOps` contract and record that public scalar `pow` remains a composite recipe rather than same-name `pto.tpows` coverage
- keep `output_normalize` and the other composite operators outside the agreed scope

## Testing

- `cmake --build build --parallel 2`
- `pre-commit run --files <changed-files>`
- focused math/reduction UT: 114 passed
- `pytest tests/st/runtime/ops/test_missing_math_ops.py --platform=a2a3 --codegen-only --forked -n 2 -q`: 56 passed
- GitHub CI on the rebased implementation commit: all required checks passed

A5 validation is intentionally out of scope.

Fixes #2665

### CODE REVIEW

File: docs/en/dev/passes/13-lower_composite_ops.md [hunk 1]
```diff
@@ -28,6 +28,28 @@ The empty `PassProperties` contract (`kLowerCompositeOpsProperties` in `include/
 
 ## Architecture
 
+### Standalone math recipes
+
+`pow`, `mean`, and `clamp` are independent tensor/tile APIs. They use FP32
+intermediates and retain FP16/FP32 input dtype. `sqrt` remains the existing primitive.
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply docs/en/dev/passes/13-lower_composite_ops.md:34]: Fixed in `d6c79d92` — updated the English/Chinese overview and structural no-op sections, plus the public C++ declaration, Python binding help text, and Python stub to list `tile.pow`, `tile.mean`, and `tile.clamp`.

File: src/ir/op/composite_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  const auto input = As<ShapedType>(base);
+  CHECK(input->shape_.size() == 2) << "mean requires rank-2 input, got rank " << input->shape_.size();
+  const int requested_axis = GetKwargOr<int>(kwargs, "axis", -1);
+  int axis = requested_axis;
+  if (axis < 0) axis += 2;
+  CHECK(axis == 0 || axis == 1) << "mean requires axis 0, 1, -1, or -2, got " << requested_axis;
+  auto valid = tile ? GetValidShape(As<TileType>(base)) : GetValidShape(AsTensorTypeLike(base));
+  CheckReductionInputNonEmpty(valid, "mean", args[0]->span_);
+  const auto count = As<ConstInt>(valid[axis]);
+  CHECK(count && count->value_ > 0) << "mean requires a positive static valid extent on the reduced axis";
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/composite_ops.cpp:93]: Resolved without an extra check: on the rebased head, `GetValidShape` calls `detail::ResolveValidShape`, which checks `valid.size() == physical.size()` before returning. `DeduceComposite` invokes it before `DeduceMean` can reach `valid[axis]`, so the proposed access is already guarded. Existing tile/tensor rank-mismatch UTs cover this path; the focused suite passed 114 tests.
